packages/runtime/container-runtime/src/dataStore.ts

```diff
@@ -0,0 +1,157 @@
+ */
+export interface IDataStoreAliasMessage {
+    /** The internal id of the datastore */
+    readonly internalId: string;
+    /** The alias name to be assigned to the datastore */
+    readonly alias: string;
+}
+
+/**
+ * Type guard that returns true if the given alias message is actually an instance of
+ * a class which implements @see IDataStoreAliasMessage
+ * @param maybeDataStoreAliasMessage - message object to be validated
+ * @returns True if the @see IDataStoreAliasMessage is fully implemented, false otherwise
+ */
+ export const isDataStoreAliasMessage = (
+    maybeDataStoreAliasMessage: any,
+): maybeDataStoreAliasMessage is IDataStoreAliasMessage => {
+    return typeof maybeDataStoreAliasMessage?.internalId === "string"
+        && typeof maybeDataStoreAliasMessage?.alias === "string";
+};
+
+/**
+ * A fluid router with the capability of being assigned an alias
+ */
+ export interface IDataStore extends IFluidRouter {
+    /**
+     * Attempt to assign an alias to the datastore.
+     * If the operation succeeds, the datastore can be referenced
+     * by the supplied alias.
+     *
+     * @param alias - Given alias for this datastore.
+     */
+    trySetAlias(alias: string): Promise<boolean>;
+}
+
+export const channelToDataStore = (
+    fluidDataStoreChannel: IFluidDataStoreChannel,
+    internalId: string,
+    runtime: ContainerRuntime,
+    logger: ITelemetryLogger,
+): IDataStore => new DataStore(fluidDataStoreChannel, internalId, runtime, logger);
+
+enum AliasState {
+    Aliased = "Aliased",
+    Aliasing = "Aliasing",
+    None = "None",
+}
+
+class DataStore implements IDataStore {
+    private aliasState: AliasState = AliasState.None;
+    private alias: string | undefined;
+
+    async trySetAlias(alias: string): Promise<boolean> {
+        assert(this.runtime.attachState !== AttachState.Detached, "Trying to alias when not attached!");
+        switch (this.aliasState) {
+            // If we're already aliasing, throw an exception
+            case AliasState.Aliasing: throw new UsageError("The datastore is already in the process of aliasing");
+            // If this datastore is already aliased, return true only if this
+            // is a repeated call for the same alias
+            case AliasState.Aliased:
+                if (this.alias !== alias) {
+                    throw new UsageError("The datastore has already been aliased");
+                }
+
+                return true;
+            // There is no current or past alias operation for this datastore,
+            // it is safe to continue execution
+            case AliasState.None: break;
+            default: unreachableCase(this.aliasState);
+        }
+
+        this.aliasState = AliasState.Aliasing;
+        this.fluidDataStoreChannel.bindToContext();
+
+        const message: IDataStoreAliasMessage = {
+            internalId: this.internalId,
+            alias,
+        };
+
+        return this.ackBasedPromise<boolean>((resolve) => {
```

Comment:
Might be worth putting comments here and there RE who handles unattached container. Looking at this code, it's not clear if this code misses this case, or it's handled in submitDataStoreAliasOp

Reply:
Ah, I see assert above. I think it should not be assert, it should be handled case that assigns alias automatically without an op. Client should not care that much about differences in these states - API should just work!